src/main/scala/ztis/VideoOrigin.scala

```diff
@@ -1,38 +1,68 @@
 package ztis
 
 sealed trait VideoOrigin {
-  def matches(host: String): Boolean
+  def matches(host: String): Option[String]
 }
 
 object VideoOrigin {
   case object Vimeo extends VideoOrigin {
-    private val domains = Set("vimeo.com", "www.vimeo.com")
+    // http://stackoverflow.com/questions/10488943/easy-way-to-get-vimeo-id-from-a-vimeo-url
 
-    override def matches(host: String): Boolean = {
-      domains.contains(host)
+    val vimeoRgx = """(?:https?://)?(?:www.)?(?:player.)?vimeo.com/(?:[a-z]*/)*([0-9]{6,11})[?]?.*"""
+                    .replaceAll("\\s*#.*\n", "").replace(" ","").r
+
+    override def matches(host: String): Option[String] = {
+      host match {
+        case vimeoRgx(a) => Some("http://vimeo.com/" + a)
+        case _ => None
+      }
     }
   }
   
   case object YouTube extends VideoOrigin {
-    private val domains = Set("youtube.com", "www.youtube.com")
-    
-    override def matches(host: String): Boolean = {
-      domains.contains(host)
+    // http://stackoverflow.com/questions/5830387/how-to-find-all-youtube-video-ids-in-a-string-using-a-regex
+    // http://stackoverflow.com/questions/11431078/scala-youtube-video-id-from-url
+
+    val youtubeRgx = """https?://         # Required scheme. Either http or https.
+                       |(?:[0-9a-zA-Z-]+\.)? # Optional subdomain.
+                       |(?:               # Group host alternatives.
+                       |  youtu\.be/      # Either youtu.be,
+                       || youtube\.com    # or youtube.com followed by
+                       |  \S*             # Allow anything up to VIDEO_ID,
+                       |  [^\w\-\s]       # but char before ID is non-ID char.
+                       |)                 # End host alternatives.
+                       |([\w\-]{11})      # $1: VIDEO_ID is exactly 11 chars.
+                       |(?=[^\w\-]|$)     # Assert next char is non-ID or EOS.
+                       |(?!               # Assert URL is not pre-linked.
+                       |  [?=&+%\w.-]*      # Allow URL (query) remainder.
+                       |  (?:             # Group pre-linked alternatives.
+                       |    [\'"][^<>]*>  # Either inside a start tag,
+                       |  | </a>          # or inside <a> element text contents.
+                       |  )               # End recognized pre-linked alts.
+                       |)                 # End negative lookahead assertion.
+                       |[?=&+%\w.-]*       # Consume any URL (query) remainder.
+                       |""".stripMargin.replaceAll("\\s*#.*\n", "").replace(" ","").r
+
+    override def matches(host: String): Option[String] = {
```

Comment:
After your changes I would change the name of the parameter - it is no longer only host part of the URL but whole URL if I understand the code correctly.